### ISSUE TITLE
Ivy TestBed should throw descriptive errors when override methods are used with incompatible types

### ISSUE THREAD

**Issue by Shijir** [angular/angular#34072]
# 🐞 bug report

### Affected Package

 The issue is caused by package "@angular/core": "9.0.0-rc.3",

### Description

When I use `TestBed.overrideComponent` to override some services with their mocks, the tests start failing with the following error:
```
 Cannot read property 'templateUrl' of null 
```

The code snippet that's causing the error:
```ts
TestBed.overrideComponent(SomeDirective, {
      set: { providers: [{ provide: SomeService, useClass: MockSomeService }] },
});
```

## 🔬 Minimal Reproduction

https://github.com/Shijir/shiny-ivy-app-test-fail

## 🔥 Exception or Error
<pre><code>Failed: Uncaught (in promise): TypeError: Cannot read property 'templateUrl' of null
TypeError: Cannot read property 'templateUrl' of null
    at componentNeedsResolution (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:21097:1)
    at maybeQueueResolutionOfComponentResources (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:21080:1)
    at compileComponent (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:39395:1)
    at http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:1941:30
    at Set.forEach (<anonymous>)
    at R3TestBedCompiler.compileTypesSync (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:1932:1)
    at R3TestBedCompiler.<anonymous> (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:1827:1)
    at Generator.next (<anonymous>)
    at http://localhost:9876/_karma_webpack_/node_modules/tslib/tslib.es6.js:73:1
    at new ZoneAwarePromise (http://localhost:9876/_karma_webpack_/node_modules/zone.js/dist/zone-evergreen.js:872:1)
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>@angular/core": "9.0.0-rc.3
</code></pre>

**Anything else relevant?**

**Comment by AndrewKushnir** [comment]
Hi @Shijir, it looks like you're using `TestBed.overrideComponent` to apply overrides to a Directive, so TestBed tries to compile `SomeDirective` as a Component. Try to use `TestBed.overrideDirective` function instead and let us know if that works as expected. Here is an example:

```
TestBed.overrideDirective(SomeDirective, {
      set: { providers: [{ provide: SomeService, useClass: MockSomeService }] },
});
```
Thank you.

**Comment by Shijir** [comment]
Hi @AndrewKushnir! Your suggestion is working great! Thank you! 

`TestBed.overrideComponent` used to work for directives in our project. But with Ivy, it looks like we should be using the correct corresponding APIs for directives, components, etc. Maybe, we could use some clear error messages in this case?

**Comment by AndrewKushnir** [comment]
@Shijir, thanks for the update. Adding more descriptive error messages makes sense, I'll work on a PR to do that. Note: I will also change the title of the ticket to reflect that. Thank you.

**Comment by AndrewKushnir** [comment]
@Shijir, PR #34204 that adds extra checks and error messages was merged into master and the changes will be available in the upcoming 9.0.0-rc.5 release. Thank you.
